File: apps/api/src/citizens/citizens.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { SupabaseService } from '../supabase/supabase.service';
import { SijilliErrors } from '../common/errors/error-envelope';
import { SijilliRequestUser } from '../auth/types';
import { CreateCitizenDto } from './dto/create-citizen.dto';
import { UpdateCitizenDto } from './dto/update-citizen.dto';
import { ListCitizensQuery } from './dto/list-citizens.dto';

const SELECT_COLS = `
  id,
  first_name_ar, father_name_ar, grandfather_name_ar, family_name_ar,
  first_name_en, father_name_en, grandfather_name_en, family_name_en,
  mother_name_ar,
  legacy_national_no, family_book_no,
  gender, birth_date, birth_place, nationality, marital_status,
  phone, email,
  region_id, municipality_id, address_ar,
  photo_path, signature_path,
  is_active, created_at, updated_at
`;

@Injectable()
export class CitizensService {
  private readonly logger = new Logger(CitizensService.name);

  constructor(private readonly supabase: SupabaseService) {}

  // ----------------------------------------------------------------------
  // Create — officer-only.
  // ----------------------------------------------------------------------
  async create(dto: CreateCitizenDto, actor: SijilliRequestUser) {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    const { data, error } = await this.supabase.admin
      .from('citizens')
      .insert({
        first_name_ar: dto.first_name_ar,
        father_name_ar: dto.father_name_ar,
        grandfather_name_ar: dto.grandfather_name_ar,
        family_name_ar: dto.family_name_ar,
        first_name_en: dto.first_name_en ?? null,
        father_name_en: dto.father_name_en ?? null,
        grandfather_name_en: dto.grandfather_name_en ?? null,
        family_name_en: dto.family_name_en ?? null,
        mother_name_ar: dto.mother_name_ar ?? null,
        legacy_national_no: dto.legacy_national_no ?? null,
        family_book_no: dto.family_book_no ?? null,
        gender: dto.gender,
        birth_date: dto.birth_date,
        birth_place: dto.birth_place ?? null,
        marital_status: dto.marital_status ?? null,
        phone: dto.phone ?? null,
        email: dto.email ?? null,
        region_id: dto.region_id,
        municipality_id: dto.municipality_id ?? null,
        address_ar: dto.address_ar ?? null,
        photo_path: dto.photo_path ?? null,
        signature_path: dto.signature_path ?? null,
        created_by: actor.officer_id,
        is_active: true,
      })
      .select(SELECT_COLS)
      .single();

    if (error || !data) {
      if (error?.code === '23505') {
        throw SijilliErrors.conflict(
          'يوجد مواطن مسجّل مسبقاً برقم وطني أو هاتف أو بريد إلكتروني مماثل.',
          'A citizen already exists with the same national/phone/email.',
        );
      }
      throw SijilliErrors.upstream(`Failed to insert citizen: ${error?.message ?? 'unknown'}`);
    }

    return data;
  }

  // ----------------------------------------------------------------------
  // List — officers see their region + (super_admin) see all.
  // Cursor pagination on created_at DESC.
  // ----------------------------------------------------------------------
  async list(q: ListCitizensQuery, actor: SijilliRequestUser) {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    let query = this.supabase.admin
      .from('citizens')
      .select(SELECT_COLS)
      .eq('is_active', true)
      .order('created_at', { ascending: false })
      .order('id', { ascending: false })
      .limit(q.limit + 1); // fetch one extra to derive next_cursor

    // Region scoping for non-super-admin roles.
    if (actor.role !== 'super_admin' && actor.role !== 'auditor') {
      if (q.region_id !== undefined && actor.region_id !== null && q.region_id !== actor.region_id) {
        throw SijilliErrors.forbidden('لا يمكنك عرض مواطنين من خارج منطقتك.');
      }
      if (actor.region_id) {
        query = query.eq('region_id', actor.region_id);
      }
    } else if (q.region_id !== undefined) {
      query = query.eq('region_id', q.region_id);
    }

    if (q.cursor) {
      query = query.lt('created_at', q.cursor);
    }

    if (q.q && q.q.trim().length >= 2) {
      // Use ilike on the concatenated name (the schema has a trigram GIN
      // index on this expression — see 003_citizens.sql).
      const safe = q.q.replace(/[%_]/g, '\\$&').trim();
      query = query.or(
        `first_name_ar.ilike.%${safe}%,family_name_ar.ilike.%${safe}%,father_name_ar.ilike.%${safe}%`,
      );
    }

    const { data, error } = await query;
    if (error) {
      throw SijilliErrors.upstream(`Failed to list citizens: ${error.message}`);
    }

    const items = data ?? [];
    let next_cursor: string | null = null;
    if (items.length > q.limit) {
      const overflow = items.pop()!;
      next_cursor = overflow.created_at as string;
    }
    return { items, next_cursor };
  }

  // ----------------------------------------------------------------------
  // Get by id — officers can read any; a citizen can read themselves.
  // ----------------------------------------------------------------------
  async getById(id: string, actor: SijilliRequestUser) {
    if (actor.role === 'citizen' && actor.citizen_id !== id) {
      throw SijilliErrors.forbidden();
    }
    const { data, error } = await this.supabase.admin
      .from('citizens')
      .select(SELECT_COLS)
      .eq('id', id)
      .maybeSingle();

    if (error) throw SijilliErrors.upstream(`Failed to load citizen: ${error.message}`);
    if (!data) throw SijilliErrors.notFound('المواطن');
    return data;
  }

  // ----------------------------------------------------------------------
  // Update — officer-only.
  // ----------------------------------------------------------------------
  async update(id: string, dto: UpdateCitizenDto, actor: SijilliRequestUser) {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    const before = await this.supabase.admin
      .from('citizens')
      .select(SELECT_COLS)
      .eq('id', id)
      .maybeSingle();

    if (before.error) throw SijilliErrors.upstream(before.error.message);
    if (!before.data) throw SijilliErrors.notFound('المواطن');

    const patch: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(dto)) {
      if (v !== undefined) patch[k] = v;
    }

    if (Object.keys(patch).length === 0) {
      return before.data;
    }

    const { data, error } = await this.supabase.admin
      .from('citizens')
      .update(patch)
      .eq('id', id)
      .select(SELECT_COLS)
      .single();

    if (error || !data) {
      throw SijilliErrors.upstream(`Failed to update citizen: ${error?.message ?? 'unknown'}`);
    }
    return data;
  }
}

File: apps/api-dotnet/Citizens/CitizensService.cs
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using Sijilli.Api.Auth;
using Sijilli.Api.Common;
using Sijilli.Api.Common.Errors;
using Sijilli.Api.Data;
using Sijilli.Api.Data.Entities;

namespace Sijilli.Api.Citizens;

public sealed class CitizensService(SijilliDbContext db)
{
    private const int UNIQUE_VIOLATION = 2627;
    private const int UNIQUE_VIOLATION_INDEX = 2601;

    public async Task<CitizenView> CreateAsync(CreateCitizenDto dto, CurrentUser actor, CancellationToken ct)
    {
        if (actor.OfficerId is null) throw SijilliException.Forbidden();

        var c = new Citizen
        {
            Id = Guid.NewGuid(),
            FirstNameAr = dto.FirstNameAr,
            FatherNameAr = dto.FatherNameAr,
            GrandfatherNameAr = dto.GrandfatherNameAr,
            FamilyNameAr = dto.FamilyNameAr,
            FirstNameEn = dto.FirstNameEn,
            FatherNameEn = dto.FatherNameEn,
            GrandfatherNameEn = dto.GrandfatherNameEn,
            FamilyNameEn = dto.FamilyNameEn,
            MotherNameAr = dto.MotherNameAr,
            LegacyNationalNo = dto.LegacyNationalNo,
            FamilyBookNo = dto.FamilyBookNo,
            Gender = dto.Gender,
            BirthDate = dto.BirthDate.ToDateTime(TimeOnly.MinValue),
            BirthPlace = dto.BirthPlace,
            MaritalStatus = dto.MaritalStatus,
            Phone = dto.Phone,
            Email = dto.Email?.ToLowerInvariant(),
            RegionId = dto.RegionId,
            MunicipalityId = dto.MunicipalityId,
            AddressAr = dto.AddressAr,
            PhotoPath = dto.PhotoPath,
            SignaturePath = dto.SignaturePath,
            CreatedBy = actor.OfficerId,
            IsActive = true,
        };

        db.Citizens.Add(c);
        try
        {
            await db.SaveChangesAsync(ct);
        }
        catch (DbUpdateException ex) when (IsUnique(ex))
        {
            throw SijilliException.Conflict(
                "يوجد مواطن مسجّل مسبقاً برقم وطني أو هاتف أو بريد إلكتروني مماثل.",
                "A citizen already exists with the same national/phone/email.");
        }
        return CitizenView.From(c);
    }

    public async Task<CursorPage<CitizenView>> ListAsync(ListCitizensQuery q, CurrentUser actor, CancellationToken ct)
    {
        if (actor.OfficerId is null) throw SijilliException.Forbidden();

        IQueryable<Citizen> query = db.Citizens.AsNoTracking().Where(c => c.IsActive);

        // Region scoping for non-super-admin / non-auditor roles.
        if (actor.Role is not ("super_admin" or "auditor"))
        {
            if (q.RegionId is not null && actor.RegionId is not null && q.RegionId != actor.RegionId)
                throw SijilliException.Forbidden("لا يمكنك عرض مواطنين من خارج منطقتك.");
            if (actor.RegionId is not null)
                query = query.Where(c => c.RegionId == actor.RegionId);
        }
        else if (q.RegionId is not null)
        {
            query = query.Where(c => c.RegionId == q.RegionId);
        }

        if (!string.IsNullOrWhiteSpace(q.Cursor) &&
            DateTimeOffset.TryParse(q.Cursor, out var cursorTs))
        {
            query = query.Where(c => c.CreatedAt < cursorTs);
        }

        if (!string.IsNullOrWhiteSpace(q.Q) && q.Q.Trim().Length >= 2)
        {
            // Arabic_CI_AS collation makes LIKE case-insensitive for Latin
            // and accent-insensitive for Arabic. Escape % and _.
            var pat = "%" + q.Q.Trim().Replace("[", "[[]").Replace("%", "[%]").Replace("_", "[_]") + "%";
            query = query.Where(c =>
                EF.Functions.Like(c.FirstNameAr, pat) ||
                EF.Functions.Like(c.FatherNameAr, pat) ||
                EF.Functions.Like(c.FamilyNameAr, pat));
        }

        var rows = await query
            .OrderByDescending(c => c.CreatedAt)
            .ThenByDescending(c => c.Id)
            .Take(q.Limit + 1)
            .ToListAsync(ct);

        string? nextCursor = null;
        if (rows.Count > q.Limit)
        {
            var overflow = rows[q.Limit];
            nextCursor = overflow.CreatedAt.ToString("o");
            rows = rows.Take(q.Limit).ToList();
        }

        return new CursorPage<CitizenView>
        {
            Items = rows.Select(CitizenView.From).ToList(),
            NextCursor = nextCursor,
        };
    }

    public async Task<CitizenView> GetByIdAsync(Guid id, CurrentUser actor, CancellationToken ct)
    {
        if (actor.Role == "citizen" && actor.CitizenId != id) throw SijilliException.Forbidden();

        var c = await db.Citizens.AsNoTracking().FirstOrDefaultAsync(x => x.Id == id, ct)
            ?? throw SijilliException.NotFound("المواطن", "Citizen");
        return CitizenView.From(c);
    }

    public async Task<CitizenView> UpdateAsync(Guid id, UpdateCitizenDto dto, CurrentUser actor, CancellationToken ct)
    {
        if (actor.OfficerId is null) throw SijilliException.Forbidden();

        var c = await db.Citizens.FirstOrDefaultAsync(x => x.Id == id, ct)
            ?? throw SijilliException.NotFound("المواطن", "Citizen");

        if (dto.FirstNameEn is not null) c.FirstNameEn = dto.FirstNameEn;
        if (dto.FatherNameEn is not null) c.FatherNameEn = dto.FatherNameEn;
        if (dto.GrandfatherNameEn is not null) c.GrandfatherNameEn = dto.GrandfatherNameEn;
        if (dto.FamilyNameEn is not null) c.FamilyNameEn = dto.FamilyNameEn;
        if (dto.MotherNameAr is not null) c.MotherNameAr = dto.MotherNameAr;
        if (dto.LegacyNationalNo is not null) c.LegacyNationalNo = dto.LegacyNationalNo;
        if (dto.FamilyBookNo is not null) c.FamilyBookNo = dto.FamilyBookNo;
        if (dto.BirthPlace is not null) c.BirthPlace = dto.BirthPlace;
        if (dto.MaritalStatus is not null) c.MaritalStatus = dto.MaritalStatus;
        if (dto.Phone is not null) c.Phone = dto.Phone;
        if (dto.Email is not null) c.Email = dto.Email.ToLowerInvariant();
        if (dto.RegionId is not null) c.RegionId = dto.RegionId;
        if (dto.MunicipalityId is not null) c.MunicipalityId = dto.MunicipalityId;
        if (dto.AddressAr is not null) c.AddressAr = dto.AddressAr;
        if (dto.PhotoPath is not null) c.PhotoPath = dto.PhotoPath;
        if (dto.SignaturePath is not null) c.SignaturePath = dto.SignaturePath;

        try
        {
            await db.SaveChangesAsync(ct);
        }
        catch (DbUpdateException ex) when (IsUnique(ex))
        {
            throw SijilliException.Conflict(
                "تعارض في رقم الهاتف أو البريد الإلكتروني.",
                "Conflict on phone or email.");
        }
        return CitizenView.From(c);
    }

    private static bool IsUnique(DbUpdateException ex) =>
        ex.InnerException is SqlException se &&
        (se.Number == UNIQUE_VIOLATION || se.Number == UNIQUE_VIOLATION_INDEX);
}
